src/engine/QueryExecutionTree.cpp

```diff
@@ -131,30 +143,35 @@ QueryExecutionTree::generateResults(const SelectedVarsOrAsterisk & selectVarsOrA
 
 template ad_utility::stream_generator::stream_generator
 QueryExecutionTree::generateResults<QueryExecutionTree::ExportSubFormat::CSV>(
-    const SelectedVarsOrAsterisk & selectVarsOrAsterisk, size_t limit, size_t offset) const;
+    const SelectedVarsOrAsterisk & selectedVarsOrAsterisk, size_t limit, size_t offset) const;
 
 template ad_utility::stream_generator::stream_generator
 QueryExecutionTree::generateResults<QueryExecutionTree::ExportSubFormat::TSV>(
-    const SelectedVarsOrAsterisk & selectVarsOrAsterisk, size_t limit, size_t offset) const;
+    const SelectedVarsOrAsterisk & selectedVarsOrAsterisk, size_t limit, size_t offset) const;
 
 template ad_utility::stream_generator::stream_generator QueryExecutionTree::
     generateResults<QueryExecutionTree::ExportSubFormat::BINARY>(
-        const SelectedVarsOrAsterisk & selectVarsOrAsterisk, size_t limit, size_t offset) const;
+        const SelectedVarsOrAsterisk & selectedVarsOrAsterisk, size_t limit, size_t offset) const;
 
 // ___________________________________________________________________________
 QueryExecutionTree::ColumnIndicesAndTypes
 QueryExecutionTree::selectedVariablesToColumnIndices(
-    const SelectedVarsOrAsterisk & selectVarsOrAsterisk,
+    SelectedVarsOrAsterisk selectedVarsOrAsterisk,
     const ResultTable& resultTable) const {
   ColumnIndicesAndTypes exportColumns;
-  if(selectVarsOrAsterisk.isAsterisk()) {
-    for(const auto& elem: getVariableColumns()) {
-      exportColumns.push_back(VariableAndColumnIndex{
-          elem.first, elem.second, resultTable.getResultType(elem.second)});
+  if(selectedVarsOrAsterisk.isAsterisk()) {
+    for(auto var: selectedVarsOrAsterisk.retrieveOrder()){
+      if (getVariableColumns().contains(var)) {
+        auto columnIndex = getVariableColumns().at(var);
+        exportColumns.push_back(VariableAndColumnIndex{
+            var, columnIndex, resultTable.getResultType(columnIndex)});
+      } else {
+        exportColumns.emplace_back(std::nullopt);
+      }
     }
```

Comment:
Also use your code from above here (including the warning etc.)
We have another PR by someone else upcoming, that will soon remove this duplication.